File: sdk/src/main/java/io/verloop/sdk/Verloop.java
package io.verloop.sdk;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.UUID;

public class Verloop {

    public static final String TAG = "VerloopOBJECT";

    static final int VERLOOP_ID = 8375667;

    static final String IS_SHOWN = "IS_ACTIVITY_ACTIVE";
    static final String ONLY_BIND = "ONLY_BIND";

    static final String CONFIG_CLIENT_ID = "CLIENT_ID";
    static final String CONFIG_USER_ID = "USER_ID";
    static final String CONFIG_FCM_TOKEN = "FCM_TOKEN";
    static final String CONFIG_STAGING = "IS_STAGING";
    static final String CONFIG_FIELDS = "CUSTOM_FIELDS";
    static final String CONFIG_RECIPE_ID = "RECIPE_ID";

    static final String CONFIG_USER_NAME = "USER_NAME";
    static final String CONFIG_USER_EMAIL = "USER_EMAIL";
    static final String CONFIG_USER_PHONE = "USER_PHONE";

    static final String SHARED_PREFERENCE_FILE_NAME = "io.verloop.sdk";

    private Context context;

    private String userId;
    private String clientId;
    private String fcmToken;
    private boolean isStaging;
    private LiveChatButtonClickListener buttonOnClickListener;
    private LiveChatUrlClickListener urlClickListener;

    /**
     * @param context Context of an activity/service.
     * @param config  The <code>VerloopConfig</code> object for the current user.
     */
    public Verloop(Context context, VerloopConfig config) {
        this.context = context;

        this.userId = retrieveUserId(config);
        this.clientId = config.getClientId();
        this.fcmToken = config.getFcmToken();
        this.isStaging = config.getStaging();
        this.buttonOnClickListener = config.getButtonOnClickListener();
        this.urlClickListener = config.getUrlClickListener();

        config.save(getPreferences());
        this.startService();
    }

    /**
     * @param userId User ID of the user you want to log in the app.
     * @deprecated Use {@link #login(VerloopConfig)} instead.
     * <p>
     * Login a different user than the one that this object was initialized with.
     */
    public void login(String userId) {
        login(userId, null);
    }

    /**
     * @param userId   User ID of the user you want to log in the app.
     * @param fcmToken FCM Token of the user being logged in.
     * @deprecated Use {@link #login(VerloopConfig)} instead.
     * <p>
     * Login a different user than the one that this object was initialized with.
     */
    public void login(String userId, String fcmToken) {
        stopService();

        this.userId = userId;
        this.fcmToken = fcmToken;

        SharedPreferences.Editor editor = getPreferences().edit();

        editor.putString(CONFIG_USER_ID, this.userId);
        editor.putString(CONFIG_FCM_TOKEN, this.fcmToken);

        editor.apply();

        startService();
    }

    /**
     * Login a different user than the one that this object was initialized with.
     *
     * @param config The <code>VerloopConfig</code> object to initialize with a new user.
     */
    public void login(VerloopConfig config) {
        this.stopService();

        this.userId = retrieveUserId(config);
        this.clientId = config.getClientId();
        this.fcmToken = config.getFcmToken();
        this.isStaging = config.getStaging();
        this.buttonOnClickListener = config.getButtonOnClickListener();
        this.urlClickListener = config.getUrlClickListener();

        config.save(getPreferences());

        this.startService();
    }

    /**
     * This will stop all the services,
     * set FCM token for user as null and
     * removes the user_id from the Shared Preferences
     */
    public void logout() {
        stopService();

        if (fcmToken != null)
            VerloopLogoutService.logout(context, clientId, userId, fcmToken, isStaging);

        SharedPreferences.Editor editor = getPreferences().edit();

        editor.putString(CONFIG_USER_ID, null);
        editor.putString(CONFIG_FCM_TOKEN, null);

        editor.apply();
    }

    /**
     * This will open up the activity for chat and load all the data provided in VerloopConfig
     */
    public void showChat() {
        startService();

        Intent i = new Intent(context, VerloopActivity.class);
        context.startActivity(i);

        if ((this.buttonOnClickListener != null || this.urlClickListener != null) && !EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }
    }

    public void hideChat() {
        EventBus.getDefault().post(new HideChatEvent());
    }

    /**
     * This method should be called if you are listening to button clicks
     * Call this in onDestroy method of the activity
     */
    public void onStopChat() {
        if ((buttonOnClickListener != null || this.urlClickListener != null) && EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this);
        }
    }

    /**
     * This method is for event listening, DO NOT call it explicitly.
     *
     * @param event
     */
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    public void onChatButtonClickEvent(ChatButtonClickEvent event) {
        if (buttonOnClickListener != null) {
            String title = event.getTitle();
            String type = event.getType();
            String payload = event.getPayload();

            Log.d(TAG, "Button click event received Title: " + title + " Type: " + type + " Payload " + payload);

            buttonOnClickListener.buttonClicked(title, type, payload);
        }
    }

    /**
     * This method is for event listening, DO NOT call it explicitly.
     *
     * @param event
     */
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    public void onChatUrlClickEvent(ChatUrlClickEvent event) {
        if (urlClickListener != null) {
            String url = event.getUrl();

            Log.d(TAG, "Url click event received Url: " + url);

            urlClickListener.urlClicked(url);
        }
    }

    private String retrieveUserId(VerloopConfig config) {
        if (config.getUserId() == null) {
            SharedPreferences preferences = getPreferences();
            userId = preferences.getString(CONFIG_USER_ID, null);

            if (userId == null) {
                userId = UUID.randomUUID().toString();
                preferences.edit().putString(CONFIG_USER_ID, userId).apply();
            }

            config.setUserId(userId);

            return userId;
        } else
            return userId = config.getUserId();
    }

    private SharedPreferences getPreferences() {
        return context.getSharedPreferences(SHARED_PREFERENCE_FILE_NAME, Context.MODE_PRIVATE);
    }

    private void startService() {
        Intent intent = new Intent(context, VerloopService.class);
        context.startService(intent);
    }

    private void stopService() {
        Intent intent = new Intent(context, VerloopService.class);
        context.stopService(intent);
    }
}

File: sdk/src/main/java/io/verloop/sdk/Verloop.kt
package io.verloop.sdk

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.Intent
import android.content.SharedPreferences
import android.util.Log
import androidx.work.*
import io.verloop.sdk.model.LogEvent
import io.verloop.sdk.model.LogoutRequestBody
import io.verloop.sdk.service.LogoutWorker
import io.verloop.sdk.ui.VerloopActivity
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class Verloop(val context: Context, var verloopConfig: VerloopConfig) {

    fun clearChat() {
        Log.d(TAG, "clearChat() called from SDK")
        val activity = io.verloop.sdk.ui.VerloopActivity.currentInstance
        if (activity != null) {
            Log.d(TAG, "VerloopActivity.currentInstance is not null")
            val fragment = activity.supportFragmentManager.findFragmentByTag("VerloopActivity#Fragment") as? io.verloop.sdk.ui.VerloopFragment
            if (fragment != null) {
                Log.d(TAG, "VerloopFragment found, calling clearChat() on fragment")
                fragment.clearChat()
                Companion.pendingCloseChat = false
            } else {
                Log.w(TAG, "VerloopFragment not found in activity, setting pendingCloseChat = true")
                Companion.pendingCloseChat = true
            }
        } else {
            Log.w(TAG, "VerloopActivity.currentInstance is null, setting pendingCloseChat = true")
            Companion.pendingCloseChat = true
        }
    }

    val TAG = "VerloopOBJECT"
    val PREF_USER_ID = "user_id"
    private val prefName = "VerloopPreference"
    private var preference: SharedPreferences = context.getSharedPreferences(prefName, MODE_PRIVATE)

    companion object {
        const val VERLOOP_ID = 8375667

        // Global scoped state for chat activity. Used by notification handled to identify is app is
        // in foreground of not
        var isActivityVisible = false

        // Centralized access to the event listeners for all Verloop objects.
        // Used for passing callback events triggered from the WebView back to the user via ViewModel
        val eventListeners = HashMap<String?, VerloopEventListener>()

    // Global flag to track if closeChat was requested before widget was ready
    @JvmStatic
    var pendingCloseChat: Boolean = false
    }

    @Deprecated("Not in use anymore")
    fun login(userId: String) {
    }

    @Deprecated("Not in use anymore")
    fun login(userId: String, fcmToken: String?) {
    }

    @Deprecated("Not in use anymore")
    fun login(verloopConfig: VerloopConfig) {
    }

    /**
     * This will open up the activity for chat and load all the data provided in VerloopConfig
     */
    fun showChat() {
        verloopConfig.userId =
            verloopConfig.userId ?: preference.getString(PREF_USER_ID, UUID.randomUUID().toString())
        preference.edit().putString(PREF_USER_ID, verloopConfig.userId).apply()

        eventListeners[verloopConfig.hashCode().toString()] = VerloopEventListener(verloopConfig)
        val i = Intent(context, VerloopActivity::class.java)
        i.putExtra("config", verloopConfig)

        // To be used as key for eventListeners map
        i.putExtra("configKey", verloopConfig.hashCode().toString())
        context.startActivity(i)
    }

    /**
     * This will logout the user and unregister the device from notification subscription.
     */
    fun logout() {
        val data = Data.Builder()
            .putString(LogoutRequestBody.CLIENT_ID, verloopConfig.clientId)
            .putString(LogoutRequestBody.USER_ID, verloopConfig.userId)
            .putString(LogoutRequestBody.FCM_TOKEN, verloopConfig.fcmToken)
            .putBoolean(LogoutRequestBody.IS_STAGING, verloopConfig.isStaging)
            .build()
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val logoutWorkRequest: WorkRequest =
            OneTimeWorkRequestBuilder<LogoutWorker>()
                .setInputData(data)
                .setConstraints(constraints)
                .build()

        WorkManager
            .getInstance(context)
            .enqueue(logoutWorkRequest)

        verloopConfig.userId = null
        verloopConfig.userName = null
        verloopConfig.userPhone = null
        verloopConfig.userEmail = null
        verloopConfig.fields = ArrayList()

        preference.edit().remove(PREF_USER_ID).apply()
    }

    @Deprecated("Not in use anymore")
    fun hideChat() {
    }

    @Deprecated("Not in use anymore. Use Logout instead")
    fun onStopChat() {
    }

    /**
     * This class is for event listening, DO NOT use it explicitly.
     */
    class VerloopEventListener internal constructor(private val config: VerloopConfig) {

        companion object {
            private const val TAG = "VerloopInterface"
        }

        @Throws(JSONException::class)
        fun onButtonClick(json: String) {
            Log.d(TAG, " onButtonClick $json")
            val jsonObject = JSONObject(json)
            val type = jsonObject.getString("type")
            val title = jsonObject.getString("title")
            val payload = jsonObject.getString("payload")
            config.buttonOnClickListener?.buttonClicked(title, type, payload);
        }

        @Throws(JSONException::class)
        fun onURLClick(json: String) {
            Log.d(TAG, " onURLClick $json")
            val jsonObject = JSONObject(json)
            val url = jsonObject.getString("url")
            config.chatUrlClickListener?.urlClicked(url)
        }

        fun onLogEvent(logEvent: LogEvent) {
            config.logEventListener?.logEvent(logEvent)
        }
    }
}